eng/common/matrix/README.md

```diff
@@ -0,0 +1,242 @@
+* [Matrix Generation behavior](#matrix-generation-behavior)
+  * [all](#all)
+  * [sparse](#sparse)
+  * [include/exclude](#includeexclude)
+  * [displayNames](#displaynames-1)
+  * [Filters](#filters)
+  * [Under the hood](#under-the-hood)
+
+
+This directory contains scripts supporting dynamic, cross-product matrix generation for azure pipeline jobs.
+It aims to replicate the [cross-product matrix functionality in github actions](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#example-running-with-more-than-one-version-of-nodejs),
+but also adds some additional features like sparse matrix generation, cross-product includes and excludes, and programmable matrix filters.
+
+This functionality is made possible by the ability for the azure pipelines yaml to take a [dynamic variable as an input
+for a job matrix definition](https://docs.microsoft.com/en-us/azure/devops/pipelines/process/phases?view=azure-devops&tabs=yaml#multi-job-configuration) (see the code sample at the bottom of the linked section).
+
+## Usage in a pipeline
+
+In order to use these scripts in a pipeline, you must provide a config file and call the matrix creation script within a powershell job.
+The job must then be marked as a dependency by a subsequent matrix job, and the matrix value pulled from a variable set by the powershell job.
+For example:
+
+```
+parameters:
+  - name: ProductMatrix
+    type: string
+    default: 'matrix.json'
+
+jobs:
+  - job: generate_matrix
+    steps:
+    - pwsh: |
+        eng/common/matrix/Create-JobMatrix.ps1 -ConfigPath ${{ parameters.ProductMatrix }}
+      name: generate_job_matrix
+      displayName: Generate Job Matrix
+
+  - job:
+    dependsOn: generate_matrix
+    strategy:
+      maxParallel: 0
+      matrix: $[ dependencies.generate_matrix.outputs['generate_job_matrix.matrix'] ]
+    steps:
+        ...
+```
+
+The generate_matrix job will log the matrix json for debugging, and run the task.setVariable vso command to add the generated matrix to the variable context.
+
+## Matrix config file syntax
+
+```
+"matrix": {
+  "<parameter1 name>": [ <values...> ],
+  "<parameter2 name>": [ <values...> ]
+}
+"include": [ <matrix>, <matrix>, ... ],
+"exclude": [ <matrix>, <matrix>, ... ],
+"displayNames": { <parameter value>: <human readable override> }
+```
+
+See `samples/matrix.json` for a full sample.
+
+### Fields
+
+#### matrix
+
+The `matrix` field defines the base cross-product matrix. The generated matrix can be full or sparse.
+
+Example:
+```
+"matrix": {
+  "operatingSystem": [
+    "windows-2019",
+    "ubuntu-18.04",
+    "macOS-10.15"
+  ],
+  "framework": [
+    "net461",
+    "netcoreapp2.1",
+    "net50"
+  ],
+  "additionalTestArguments": [
+    "",
+    "/p:UseProjectReferenceToAzureClients=true",
+  ]
+}
+```
+
+#### include
+
+The `include` field defines any number of matrices to be appended to the base matrix after processing exclusions.
```

Comment:
Would it be after processing exclusions? I would expect the higher level filter to potentially exclude this include as well.

Reply:
I based the behavior off the [github actions behavior](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#example-excluding-configurations-from-a-matrix). I think the reasoning is that you could do a broad exclude, but then need to add something back. Once we set the behavior we'll want to stick to it, so maybe we should brainstorm some scenarios. We could also support exclude/include priority but that feels like an overcomplication (which is ironic to say given what this PR is...).

Reply:
I was thinking about the scenario of I want to run only one entry from the matrix for manual testing. So I was expecting a constraint that runs after the full matrix processing is done.